addons/web_editor/static/src/js/editor/snippets.editor.js

```diff
@@ -1902,17 +1902,57 @@ var SnippetsMenu = Widget.extend({
             const $invisibleSnippets = globalSelector.all().find('.o_snippet_invisible').addBack('.o_snippet_invisible');
 
             $invisibleDOMPanelEl.toggleClass('d-none', !$invisibleSnippets.length);
-
-            const proms = _.map($invisibleSnippets, async el => {
-                const editor = await this._createSnippetEditor($(el));
-                const $invisEntry = $('<div/>', {
-                    class: 'o_we_invisible_entry d-flex align-items-center justify-content-between',
-                    text: editor.getName(),
-                }).append($('<i/>', {class: `fa ${editor.isTargetVisible() ? 'fa-eye' : 'fa-eye-slash'} ml-2`}));
-                $invisibleDOMPanelEl.append($invisEntry);
-                this.invisibleDOMMap.set($invisEntry[0], el);
+            // descendantPerSnippet: a map with its keys set to invisible
+            // snippets that have invisible descendants. The value corresponding
+            // to an invisible snippet element is a list filled with all its
+            // descendant invisible snippets except those that have a closer
+            // invisible snippet ancestor.
+            const descendantPerSnippet = new Map();
+            // Filter the $invisibleSnippets to only keep the root snippets and
+            // create the map (descendantPerSnippet) of the snippets and their
+            // descendant snippets.
+            const rootInvisibleSnippetEls = [...$invisibleSnippets].filter(invisibleSnippetEl => {
+                const ancestorInvisibleElement = invisibleSnippetEl.parentElement.closest(".o_snippet_invisible");
+                if (ancestorInvisibleElement) {
+                    const descendantSnippets = descendantPerSnippet.get(ancestorInvisibleElement) || [];
+                    descendantPerSnippet.set(ancestorInvisibleElement, [...descendantSnippets, invisibleSnippetEl]);
+                    return false;
+                }
+                return true;
             });
-            return Promise.all(proms);
+            // Insert an invisible snippet in its parentEl element.
+            const createInvisibleElement = async (invisibleSnippetEl, isDescendant, hasDescendant, parentEl) => {
+                const editor = await this._createSnippetEditor($(invisibleSnippetEl));
+                const invisibleEntryEl = document.createElement("div");
+                invisibleEntryEl.className = `o_we_invisible_entry d-flex align-items-center justify-content-between ${isDescendant || hasDescendant ? "o_we_invisible_element_descendant" : ""}`;
+                invisibleEntryEl.innerText = isDescendant ? "└ " + editor.getName() : editor.getName();
+                invisibleEntryEl.insertAdjacentHTML("beforeend", `<i class = "fa ${editor.isTargetVisible() ? "fa-eye" : "fa-eye-slash"} ml-2"></i>`);
```

Comment:
I'm surprised that `insertAdjacentHTML` doesn't alert CI security... It seems to me that this causes the same problems as `.innerHTML` :thinking: . I think it's better to use something like `insertAdjacentElement` or `appendChild`